Reusing  logic with custom hooks 1 (#45)
LINE-NEXT-LEVEL/client: docs/learn/Escape Hatches/reusing-logic-with-custom-hooks.mdx

import { Code } from '../../../src/components/Code';

# 커스텀 훅으로 로직 재사용하기

React는 `useState`, `useContext`, `useEffect`와 같은 내장 훅들을 갖고 있습니다.
때때로, 여러분은 더 구체적인 목적을 가진 훅이 있었으면 할 것 입니다: 예를 들어 데이터를 가져오기 위한 것,
유저가 온라인 상태인지를 지속적으로 확인하는 것, 채팅방에 연결하는 것 입니다. 이 훅들을 React에서 찾을 순 없지만,
어플리케이션의 요구사항에 따라 여러분의 훅을 작성할 수 있습니다.

:::info 여러분이 배울 것

- 커스텀 훅이 무엇인지와 어떻게 작성할 수 있는지
- 컴포넌트간의 로직을 재사용 하는 법
- 커스텀 훅의 이름과 구조를 정하는 법
- 커스텀 훅을 왜, 언제 추출하는지

:::

여러분이 네트워크에 크게 의존하는 앱(대부분의 앱이 그렇듯이)을 개발하고 있다고 상상해봅시다.
여러분이 유저들이 앱을 사용하는 중에 그들의 네트워크 연결이 우연히 끊기면 경고를 보내고 싶습니다.

> 어떻게 하시겠습니까? 여러분의 컴포넌트에는 두가지가 필요할 것으로 보입니다.

Reviewer: ```suggestion
어떻게 하시겠습니까? 여러분의 컴포넌트에는 두 가지가 필요할 것으로 보입니다.
```

1. 네트워크 온라인 여부를 추적하는 상태 하나

> 2. 전역적으로 [`online`](https://developer.mozilla.org/en-US/docs/Web/API/Window/online_event)과 [`offline`](https://developer.mozilla.org/en-US/docs/Web/API/Window/offline_event)이벤트들 구독하고 상태를 업데이트하는 Effect
>    이걸로 여러분의 컴포넌트를 네트워크 상태와 [동기화 되도록](https://react.dev/learn/synchronizing-with-effects)할 수 있습니다.
>    아래와 같이 시작할 수 있습니다.

Reviewer: ```suggestion
2. 전역적으로 [`online`](https://developer.mozilla.org/en-US/docs/Web/API/Window/online_event)과 [`offline`](https://developer.mozilla.org/en-US/docs/Web/API/Window/offline_event)이벤트들 구독하고 상태를 업데이트하는 Effect

   이걸로 여러분의 컴포넌트를 네트워크 상태와 [동기화 되도록](https://react.dev/learn/synchronizing-with-effects)할 수 있습니다.
   아래와 같이 시작할 수 있습니다.
```

<Code src="https://codesandbox.io/s/nsny83?file=%2FApp.js&utm_medium=sandpack" />

네트워크를 켜고 꺼보시고, 이 `StatusBar`가 여러분의 행동에 반응하여 어떻게 업데이트 되는지 살펴보세요.

> 또한 여러분이 다른 컴포넌트에서 같은 로직을 사용하고 싶다고 상상해봅시다.

Reviewer: ```suggestion
*또한* 여러분이 다른 컴포넌트에서 같은 로직을 사용하고 싶다고 상상해봅시다.
```

```js
function StatusBar() {
  const isOnline = useOnlineStatus();
  return <h1>{isOnline ? '✅ Online' : '❌ Disconnected'}</h1>;
}

function SaveButton() {
  const isOnline = useOnlineStatus();

  function handleSaveClick() {
    console.log('✅ Progress saved');
  }

  return (
    <button disabled={!isOnline} onClick={handleSaveClick}>
      {isOnline ? 'Save progress' : 'Reconnecting...'}
    </button>
  );
}
```

비록 그런 내장 훅은 없었지만, 여러분이 직접 작성할 수 있습니다. `useOnlineStatus`라는 함수를 선언하고, 여러분이 이전에 작성한 컴포넌트에서 모든 중복된 코드를 옮겨보세요.

```js
function useOnlineStatus() {
  const [isOnline, setIsOnline] = useState(true);
  useEffect(() => {
    function handleOnline() {
      setIsOnline(true);
    }
    function handleOffline() {
      setIsOnline(false);
    }
    window.addEventListener('online', handleOnline);
    window.addEventListener('offline', handleOffline);
    return () => {
      window.removeEventListener('online', handleOnline);
      window.removeEventListener('offline', handleOffline);
    };
  }, []);
  return isOnline;
}
```

> 함수의 마지막에는 `isOnline`을 리턴하세요. 이는 여러분의 컴포넌트가 그 값을 읽도록 합니다.

Reviewer: ```suggestion
함수의 마지막에는 `isOnline`을 리턴하세요. 이는 여러분의 컴포넌트가 그 값을 읽도록 합니다:
```

## Hook 이름은 언제나 `use`로 시작합니다.

React 어플리케이션은 컴포넌트로 이루어져있습니다.
컴포넌트는 내장, 커스텀에 상관없이 Hooks로 작성됩니다.
여러분은 대개 다른사람들이 만든 커스텀 Hooks를 사용할 것이지만, 때로는 직접 작성하기도 할 것 입니다.

여러분은 다음의 네이밍 컨벤션들을 따라야합니다:

1. **React 컴포넌트 이름은 `StatusBar`, `SaveButton`과 같이 대문자로 시작되어야합니다.** React 컴포넌트는 또한 JSX처럼 React가
   표현할줄 아는 것들을 리턴해야합니다.
2. **Hook 이름은`[useState](https://react.dev/reference/react/useState)`(내장) 또는 `useOnlineStatus`(커스텀, 이 페이지에서처럼) 와 같이 `use`로 시작하고 대문자가 따라와야합니다.** Hook은 임의의 값을 반환합니다.

이 컨벤션은 여러분이 컴포넌트를 볼 때, 상태, Effects 그리고
“숨겨진” React 기능들이 어디에 있는지 알 수 있게 합니다.
예를 들어 여러분이 `getColor()` 함수를 컴포넌트 내부에서 본다면,
그 것이 `use`로 시작되지 않기 때문에 React 상태를 포함하고 있을 가능성이 없다는 것을 확신할 수 있습니다. 하지만 `useOnlineStatus()`와 같이 불리는 함수는 대부분 내부에 다른 Hooks에 대한 호출을 포함할 것 입니다.

:::info Note
여러분의 린터가 [React를 위해 구성](https://react.dev/learn/editor-setup#linting)되었다면,
이 네이밍 컨벤션을 강조할 것 입니다. 위의 샌드박스로 스크롤을 올려보고 `useOnlineStatus`를
`getOnlineStatus`로 변경해보세요. 린터가 그 내부에 `useState`나 `useEffect`를 더 이상 호출하지 못하도록함을 확인하세요.
오직 Hooks와 컴포넌트들만이 다른 Hooks를 호출할 수 있습니다!
:::

:::note Deep Dive

## 렌더링에 호출되는 모든 함수들은 use 접두사로 시작되어야 하나요?

아닙니다. Hook을 *호출*하지 않는 함수들은 Hooks가 _될_ 필요가 없습니다.

여러분의 함수가 Hooks를 호출하지 않는다면, `use` 접두사를 피하세요. 대신에 `use` 접두사 _없이_ 일반적인 접두사로 작성하세요.

> 예를들어 아래의 `useSorted`는 Hooks를 호출하지 않으므로, 대신 `getSorted`라고 명명하세요.

Reviewer: ```suggestion
예를들어 아래의 `useSorted`는 Hooks를 호출하지 않으므로, 대신 `getSorted`라고 명명하세요:
```

```js

> // 🔴 Avoid: A Hook that doesn't use Hooks

Reviewer: 주석도 해석해주세요!

function useSorted(items) {
  return items.slice().sort();
}

> // ✅ Good: A regular function that doesn't use Hooks

Reviewer: 주석도 해석해주세요!

function getSorted(items) {
  return items.slice().sort();
}
```

> 이로서 여러분은 이 일반 함수를 조건들을 포함하여 어디서나 호출할 수 있음을 보장할 수 있습니다.

Reviewer: ```suggestion
이로서 여러분은 이 일반 함수를 조건들을 포함하여 어디서나 호출할 수 있음을 보장할 수 있습니다:
```

```js
function List({ items, shouldSort }) {
  let displayedItems = items;
  if (shouldSort) {

>     // ✅ It's ok to call getSorted() conditionally because it's not a Hook

Reviewer: 주석도 해석해주세요!

Author: 주석 해석되었습니다. [a097058](https://github.com/LINE-NEXT-LEVEL/client/pull/45/commits/a0970585645862f0ffbe27a20d7388757d12e5bd)

    displayedItems = getSorted(items);
  }
  // ...
}
```

내부에 최소 하나의 Hook을 사용하는 함수에는 `use` 접두사를 부여하세요(그렇게 Hook을 만드세요):

```js
// ✅ Good: A Hook that uses other Hooks
function useAuth() {
  return useContext(Auth);
}
```

기술적으로, React에 의해 강제되는 사항은 아닙니다. 원리적으로 다른 Hook을 부르지 않는 Hook을 만들 수 있습니다.
이는 종종 혼란스럽거나 제한적이기 때문에 그 패턴을 피하는 것이 최선입니다. 하지만 도움이 되는 흔하지 않은 케이스들이 있기도 합니다.
예를 들어 여러분의 함수가 어떠한 Hook을 사용하지 않지만, 미래에 Hook 호출들을 추가할 계획이 있는 경우입니다.

> 그런 경우, `use` 접두사를 붙이는 것이 합리적입니다.

Reviewer: ```suggestion
그런 경우, `use` 접두사를 붙이는 것이 합리적입니다:
```

    roomId: roomId,
    serverUrl: serverUrl,
  });

  return (
    <>
      <label>
        Server URL:
        <input value={serverUrl} onChange={(e) => setServerUrl(e.target.value)} />
      </label>
      <h1>Welcome to the {roomId} room!</h1>
    </>
  );
}
```

이게 훨씬 간단해 보이네요!(같은 것이지만요)
로직이 prop과 상태 변경들에 *여전히 반응하는 것*을 확인하세요.
서버 URL을 변경하거나 선택된 방을 수정해보세요:

<Code src="https://codesandbox.io/s/2bjiw9?file=%2FChatRoom.js&utm_medium=sandpack" />

하나의 Hook의 반환 값을 어떻게 취하는지 확인하세요.

```js
export default function ChatRoom({ roomId }) {
  const [serverUrl, setServerUrl] = useState('https://localhost:1234');

  useChatRoom({
    roomId: roomId,
    serverUrl: serverUrl
  });
  // ...
```

그리고 다른 Hook의 입력 값으로 넘겨주세요.

```js
export default function ChatRoom({ roomId }) {
  const [serverUrl, setServerUrl] = useState('https://localhost:1234');

  useChatRoom({
    roomId: roomId,
    serverUrl: serverUrl
  });
  // ...
```

`ChatRoom` 컴포넌트가 재렌더링 될 때마다, 그것은 최신의 `roomId`와 serverUrl`을 여러분의 Hook에 넘겨줍니다. 여러분의 Effect가 리렌더링 후에 값들이 다를 때 재연결하는 이유입니다. (만약 음향이나 비디오 프로세싱 소프트웨어로 작업해보았다면, 이와 같이 Hooks를 연결하는 것은 시각 또는 음향 효과들을 연결하는 걸 떠오르게 할 것입니다. `useState`의 결과 값이 `useChatRoom`의 입력 값으로 “먹여 주는 것”과 같습니다.)

> \*\*커스텀 Hooks로 이벤트 핸들러 넘기기

Reviewer 2: ```suggestion
## 커스텀 Hooks로 이벤트 핸들러 넘기기
```

      roomId: roomId,
    };
    const connection = createConnection(options);
    connection.connect();
    connection.on('message', (msg) => {
      showNotification('New message: ' + msg);
    });
    return () => connection.disconnect();
  }, [roomId, serverUrl]);
}
```

여러분이 컴포넌트에 로직을 되돌리길 원하다고 가정합시다.

```js
export default function ChatRoom({ roomId }) {
  const [serverUrl, setServerUrl] = useState('https://localhost:1234');

  useChatRoom({
    roomId: roomId,
    serverUrl: serverUrl,
    onReceiveMessage(msg) {
      showNotification('New message: ' + msg);
    }
  });
  // ...
```

동작하게 하기 위해서는, 여러분의 커스텀 Hook이 명명된 옵션중 하나로 `onReceiveMessage`를 받도록 변경하세요.

```js
export function useChatRoom({ serverUrl, roomId, onReceiveMessage }) {
  useEffect(() => {
    const options = {
      serverUrl: serverUrl,
      roomId: roomId,
    };
    const connection = createConnection(options);
    connection.connect();
    connection.on('message', (msg) => {
      onReceiveMessage(msg);
    });
    return () => connection.disconnect();
  }, [roomId, serverUrl, onReceiveMessage]); // ✅ All dependencies declared
}
```

동작은 하겠지만, 여러분의 커스텀 Hook이 이벤트 핸들러들을 받을 때 할 수 있는 하나의 개선이 더 있습니다.

> `onReceiveMessage` 종속성을 추가하는 것은 컴포넌트 재랜더링마다 채팅 재연결을 만들기 때문에 이상적이지 않습니다.

Reviewer 2: ```suggestion
`onReceiveMessage` 종속성을 추가하는 것은 컴포넌트를 다시 렌더링 할때마다 채팅 연결을 다시 하기 때문에 이상적이지 않습니다.
```

[이 이벤트 핸들러를 Effect Event로 감싸서 종속성에서 제거하세요](https://react.dev/learn/removing-effect-dependencies#wrapping-an-event-handler-from-the-props)

```js
import { useEffect, useEffectEvent } from 'react';
// ...

export function useChatRoom({ serverUrl, roomId, onReceiveMessage }) {
  const onMessage = useEffectEvent(onReceiveMessage);

  useEffect(() => {
    const options = {
      serverUrl: serverUrl,
      roomId: roomId,
    };
    const connection = createConnection(options);
    connection.connect();
    connection.on('message', (msg) => {
      onMessage(msg);
    });
    return () => connection.disconnect();

>   }, [roomId, serverUrl]); // ✅ All dependencies declared

Reviewer 2: 주서쿠 번여쿠 오네가이시마스

}
```

이제 채팅은 `ChatRoom` 컴포넌트가 리랜더링될 때마다 재연결되지 않습니다.
여러분이 갖고 놀 수 있는 커스텀 Hook으로 이벤트 핸들러를 넘기는 완전한 데모가 있습니다.

<Code src="https://codesandbox.io/s/3bzejk?file=%2FChatRoom.js&utm_medium=sandpack" />

이제 `useChatRoom`을 사용하기 위해 _어떻게_ 동작하는지 알 필요가 없음을 확인하세요.
다른 어떤 컴포넌트에도 추가할 수 있고, 어떤 다른 옵션들도 넘겨도 같은 방식으로 동작할 것 입니다.
이 것이 커스텀 Hook의 힘입니다.

## 커스텀 Hook을 사용 해야할 때

작은 중복 코드마다 커스텀 Hook으로 추출할 필요는 없습니다. 일부 중복은 괜찮습니다.
예를 들어, 이전처럼 하나의 `useState` 호출을 감싸기 위해 `useFormInput` Hook을 추출하는 것은 아마도 불필요할 것입니다.

하지만, Effect를 작성할때마다, 커스텀 Hook으로 감싸는 것이 더 명확한지를 고려하세요.
[Effects를 매우 자주 필요로 하면 안되기 때문에](https://react.dev/learn/you-might-not-need-an-effect), 여러분이 하나를 작성할 때,
어떤 외부 시스템이나 React가 내장 API로 지원하지 않는 어떤 것과 동기화 하기위해 "React 외부로 나가야"할 필요가 있다는 의미입니다.
커스텀 Hooks로 감싸는 것은 여러분의 의도와 어떤 데이터 흐름을 가지는지 명확히 소통하도록 해줍니다.

> 예를 들어 두 드롭다운을 표현하는 `shippingForm` 컴포넌트를 생각해보세요: 하나는 도시들의 목록을 보여주고,

Reviewer 2: ```suggestion
예를 들어 두 드롭다운을 표현하는 `shippingForm` 컴포넌트를 생각해보세요. 하나는 도시들의 목록을 보여주고,
```

다른 하나는 선택된 도시의 지역들의 목록을 보여줍니다. 아래와 같은 코드로 시작하게 될 것입니다.

```js
function ShippingForm({ country }) {
  const [cities, setCities] = useState(null);
  // This Effect fetches cities for a country
  useEffect(() => {
    let ignore = false;
    fetch(`/api/cities?country=${country}`)
      .then(response => response.json())
      .then(json => {
        if (!ignore) {
          setCities(json);
        }
      });
    return () => {
      ignore = true;
    };
  }, [country]);

  const [city, setCity] = useState(null);
  const [areas, setAreas] = useState(null);
  // This Effect fetches areas for the selected city
  useEffect(() => {
    if (city) {
      let ignore = false;
      fetch(`/api/areas?city=${city}`)
        .then(response => response.json())
        .then(json => {
          if (!ignore) {
            setAreas(json);
          }
        });
      return () => {
        ignore = true;
      };
    }
  }, [city]);

  // ...
```

> 비록 이 코드들은 매우 반복적이지만, [Effects를 각각 분리해두는 것은 옳습니다.](https://react.dev/learn/removing-effect-dependencies#is-your-effect-doing-several-unrelated-things)

Reviewer 2: ```suggestion
비록 이 코드들은 매우 반복적이지만, [Effects를 각각 분리해두는 게 좋습니다.](https://react.dev/learn/removing-effect-dependencies#is-your-effect-doing-several-unrelated-things)
```

다른 두가지를 동기화하므로, 하나의 Effect로 합치면 안됩니다. 대신에 여러분은 `shippingForm` 컴포넌트를 공통되는 로직을

> `useData` Hook으로 추출하여서 간단하게 만들 수 있습니다.

Reviewer 2: ```suggestion
`useData` Hook으로 추출해서 간단하게 만들 수 있습니다.
```
